include/mscclpp/fifo_device.hpp

```diff
@@ -64,9 +64,10 @@ struct FifoDeviceHandle {
 
     ProxyTrigger* triggerPtr = &(this->triggers[curFifoHead % size]);
 
-    // store with memory order release so that the while loop does not go pass this.
+    // There is a Write-After-Read hazard for the triggerPtr->fst. So the st instruction will not be executed
+    // before the loop.
 #if defined(MSCCLPP_DEVICE_CUDA)
-    asm volatile("st.global.release.sys.v2.u64 [%0], {%1,%2};" ::"l"(triggerPtr), "l"(trigger.fst), "l"(trigger.snd));
+    asm volatile("st.volatile.global.v2.u64 [%0], {%1,%2};" ::"l"(triggerPtr), "l"(trigger.fst), "l"(trigger.snd));
```

Comment:
I think this is dangerous since the polling is with `memoryOrderRelaxed`. So, it is possible that the ptx write happens before the poll is done. So, I suggest we change the memory order for polling to release. It should not matter much given that we should rarely get into like 60 when fifo size is large enough.

Reply:
After synced offline, I think we agree using `atomic_store` with `memory_order_relax` is safe in this case. Feel free to add comments if have other concerns